### PR TITLE
x64Emitter: Generify ABI_CallFunction variants

### PR DESCRIPTION
**Pull request by lioncash**
Removes the need for casting to void at every usage.

Note that I had to move around the register enum in order to break cyclic inclusions from `x64ABI.h`. If there's a more preferable way to reorganize this, I'm open to suggestions.

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4166)

### CODE REVIEW

File: Source/Core/Common/x64Emitter.h [hunk 1]
```diff
@@ -1090,29 +1005,148 @@ class XEmitter
...
+      CALLptr(R(RAX));
+    }
+    else
+    {
+      CALL(ptr);
+    }
+  }
+
+  template <typename FunctionPointer>
+  void ABI_CallFunctionC16(FunctionPointer func, u16 param1)
```
Comment (degasus) [inline Source/Core/Common/x64Emitter.h]: What do you think about a generic
template<typename T> OpArg TooImm(T param)
function to hide all kind of different parameters? We likely need it for u8, u16, u32, u64, and void*.

**Comment by degasus** [comment]
else, LGTM. But I'm very bad at reviewing X64 code.

File: Source/Core/Common/x64Emitter.h (hunk 1 above)
Comment (lioncash) [inline Source/Core/Common/x64Emitter.h]: I can try something like that as a follow-up, I suppose. I was mainly trying to get rid of all the casting related to different function types.

**Comment by phire** [comment]
:lgtm:

---

Reviewed 17 of 18 files at r1.
Review status: 17 of 18 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/Common/x64Emitter.h, line 1032 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4166#-KQYctS8J5uMVxr-9_he-r1-1032:-KQggvXiauugY4Q3h4hj:b-hg0x70) ([raw file](https://github.com/dolphin-emu/dolphin/blob/13506d3c12c49b1f0456fce6a985068bdfae8e48/Source/Core/Common/x64Emitter.h#L1032)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> I can try something like that as a follow-up, I suppose. I was mainly trying to get rid of all the casting related to different function types.
</details>
Sounds Good

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4166#-:-KQghW8b7NGB3LE6L3WO:bnfp4nl)*

**Pull request merged**